docs/design: Graceful Exit (#2866)
storj/storj: docs/design/storagenode-graceful-exit-protocol.md

Transferring a piece to another node may fail, hence we need to ensure that critical pieces get transferred. Storage nodes can be malicious and try to misreport transfer as "failed" or "completed". A storage node may also try to send incorrect data. This means we need proof that the correct piece was transferred.

After all pieces have been transferred the storage node needs a receipt for completing the transfer.

Both storage node and satellite operators need insight into the Graceful Exit progress.

## Design

The storage node has a Graceful Exit service, which ensures that the process is completed. It queries `graceful_exit_status` to find unfinished exits and starts a per satellite `worker` if needed.

The `worker` polls the satellite, requesting pieces to transfer. The satellite will initiate a Graceful Exit if not already initiated. When first initiated, the satellite will start gathering pieces for the exiting node and return `NotReady`. The satellite will continue to return `NotReady` until the piece gathering process has completed.

The `worker` should continue to poll the satellite at a configurable interval until it returns pieces to transfer.  

> The storage node should concurrently transfer pieces returned by the satellite. The storage node should send a `TransferSucceeded` message as pieces are successfuly transfered. The Storage node should send a `TransferFailed`, with reason, on failure.

Author: Depending on the failure, it will get reprocessed when the satellite iterates over the transfer queue.

The satellites should set the `finished_at` on success, and respond with a `DeletePiece` message. Otherwise set the `failed_at` and `failure_status_code` for reprocessing.

The satellite should respond with a `ExitCompleted` message when all pieces have finished processing. 
The storage node should store the completion receipt, stop transfer processing, and remove all remaining pieces for the satellite.

> When the storage node has failed too many transfers or has sent incorrect data, the satellite will send `ExitFailed`. This indicates that the process has ended ungracefully.

Reviewer: I think the satellite should send a completion receipt in this case too, with failed status and reason for the failure.

Author: Updated

### Transferring a Piece

The satellite picks a piece from the transfer queue and verifies whether it still needs to be transferred. During gathering or the Graceful Exit process the segment may have been changed or deleted.

The satellite will send a `TransferPiece`, which contains sufficient information to upload the piece to a replacement node.

The storage node will start a new piece upload to the replacement node similar to uplink. It will use `uplink/piecestore.Upload`. Once uploaded, the storage node will verify the piece hash sent by the replacement node corresponds to the piece hash stored in the database.

#### Verifying transfer

The storage node sends the piece hash, order limit, original "uploader" signature, and piece hash signed by the replacement node to the satellite for confirmation.

> The satellite verifies that the original piece hash matches the replacement piece hash, and verifies the order limits signature. On success, the satellite will update segment / pointer information. If verification fails, the satellite will send `ExitFailed`.

Author: Updated.  On failed exit, the repairer will correct the pointers.

> ![Transfer Sequence](./images/storagenode-graceful-exit-transfer-sequence.svg)

Reviewer 2: where is this image?

> ![Transfer Sequence](./images/storagenode-graceful-exit-transfer-sequence.svg)

Author: Updated

#### Updating the Segment / Pointer

> When the upload has been verified the satellite will use `CompareAndSwap` to only switch the appropriate node without changing others. During the upload there may have been other things that happened:

Reviewer 2: I find this section a bit difficult to follow. Am I understanding correctly that these list items are describing how we use compare-and-swap and that it's the node ID in the segment/pointer that we're comparing-and-swapping?

>Segment / pointer was reuploaded. We'll ignore the newly uploaded piece.

Is this the scenario you're describing and/or is this how re-upload would work?
```
client     "re-upload" storagenode    "transfer" storagenode    "original" storagenode           satellite
   |                 |                          |                         |                          |
   |                 |                          |     transfer piece      |                          |
   |                 |                          |            /------------A                          |
   |                 |                          B-----------/             |                          |
   |                 |                          |                         |                          |
   |                 |                          |      piece hash         |                          |
   |                 |                          |-----------\             |                          |
   |                 |                          |            \------------|                          |
   |                 |                          |                         |                          |
   |                 |                          |                         | begin piece verification |
   |                 |                          |                         |------------\             |
   |                 |                          |                         |             \------------C
   |                 |                          |                         |                          |
   |                 |                          |                         |                          |
 . |.  .  .  .  .  . |.  .  .  .  .  .  .  .  . |.  .  .  .  .  .  .  .  .| .  .  .  .  .  .  .  .  .| .
   |                 |                          |                         |    begin re-upload       |
 . |-----------------------------------------------------------------------------------\             | .
   |                 |                          |                         |             \------------D
 . |                 |                          |                         |                          | .
   |/\/\/\/\/\/\/\/\/|\/\/\/\/\/\/\/\/\/\/\/\/\/|\/\/\/\/\/\/\/\/\/\/\/\/\|/\/\/\/\/\/\/\/\/\/\/\/\/\|
 . |                 |                          |                         |                          | .
   | re-upload piece |                          |                         |                          |
 . |--------\        |                          |                         |                          | .
   |         \-------E                          |                         |                          |
 . |                 |                          |                         |                          | .
   |/\/\/\/\/\/\/\/\/|\/\/\/\/\/\/\/\/\/\/\/\/\/|\/\/\/\/\/\/\/\/\/\/\/\/\|/\/\/\/\/\/\/\/\/\/\/\/\/\|
 . |.  .  .  .  .  . |.  .  .  .  .  .  .  .  . |.  .  .  .  .  .  .  .  .| .  .  .  .  .  .  .  .  .| .
   |                 |                          |                         |                          |
   |                 |                          |                         |                          |
   |                 |                          |                         | verification complete    |
   |                 |                          |                         |             /------------|
   |                 |                          |                         F------------/             |
```

Is "newly uploaded piece" the one that was uploaded to the "transfer" node at `B` or the "re-upload" node at `E`?

> Segment / pointer pieces were changed, in that case we load the new one, check whether the exiting storage node ID still needs to be replaced. If it does, then we retry `CompareAndSwap`. When the storage node at that piece number has changed, we'll ignore the newly uploaded piece, because one of them needs to be discarded.

Again, "new one" is a bit ambiguous in my opinion. I'm assuming that since is's possible for the exiting node to still need replacement but then I wasn't able to follow the rest of the logic.

>Segment / pointer was deleted or is being deleted. We'll ignore the newly uploaded piece.
Same question.

Maybe it would be easier to put these in terms of whether we will or won't swap with the transfer node.

- Segment / pointer was reuploaded. We'll ignore piece transferred for the exit.

> - Segment / pointer pieces were changed, in that case we load the pointer, check whether the exiting storage node ID still needs to be replaced. If it does, then we retry `CompareAndSwap`. When the storage node at that piece number has changed, we'll ignore the piece transferred for the exit, because one of them needs to be discarded.

Reviewer 2: ```suggestion
- Segment / pointer pieces were changed, in that case we load the pointer and retry `CompareAndSwap`. When the storage node at that piece number has changed, we'll ignore the piece transferred for the exit, because one of them needs to be discarded.
```

Reviewer 2: Isn't that the same thing?

- Segment / pointer was deleted or is being deleted. We'll ignore the piece transferred for the exit.

> After changing the segment / pointer, we'll delete the piece from transfer queue, update progress, and send a `DeletePiece` to the exiting node.

Author: As discussed, leaving as is.

#### Handling transfer failure

> If a storage node report a failure to the satellite if it fails to upload to the replacement node, or the replacement node returns a bad piece hash. The satellite will track the failures and retry the transfer to another node.

Reviewer: Hmm, not really... My understanding is that the satellite will send `ExitFailed` if the exiting nodes send the satellite info that looks like cheating. Here we are talking the transfer may fail because the replacement node returns a hash that the exiting node does not agree to.

> If a storage node report a failure to the satellite if it fails to upload to the replacement node, or the replacement node returns a bad piece hash. The satellite will track the failures and retry the transfer to another node.

Reviewer 3: Here, the exiting node should report it as "transfer failed". The other part is about exiting node sending a piece hash which is invalid.

3. Implement transferring a single piece on storage node.
4. Implement swapping node in segment.
5. Implement non-async protocol for exiting.
6. Implement async protocol for exiting.

### Sketch

```
### storagenode ###

for {
    stream := dial satellite

    for {
        msg := stream.Recv(&msg)
        if msg is NotReady {
            go sleep a bit
            and retry later
        }

        if msg is Completed {
            update database about completion
            delete all pieces
            exit
        }

        if msg is TransferConfirmed {
            update the progress table
            delete the piece from disk
            exit
        }

        if msg is TransferPiece {
            // transfer multiple pieces in parallel, configurable
            limiter.Go(func(){
                result := try transfer msg.PieceID
                stream.Send(result)
                update progress table
            })
        }
    }
}

### satellite ###

if !initiated {
    then add node to the gracefully exiting list
    send NotReady
    return
}

if !pieces collected {
    send NotReady
    return
}

inprogress pieces
more pieces := true

> when storage node prematurely exits

Reviewer: How does the satellite know if the storage node exited prematurely?

Reviewer 3: It gets io.EOF when trying to read the next message.

    update metainfo database with the new storage node
    delete from inprogress

    stream.Send TransferConfirmed
}

stream.Send Completed with receipt
```

``` proto
service GracefulExit {
    rpc Process(stream StorageNodeMessage) returns (stream SatelliteMessage)
}

message StorageNodeMessage {
    oneof Message {
        message TransferSucceeded {
            AddressedOrderLimit addressed_order_limit;
            PieceHash original_piece_hash;
            PieceHash replacement_piece_hash;
        }
        message TransferFailed {
            enum Error {
                NOT_FOUND = 0;
                STORAGE_NODE_UNAVAILABLE = 1;
                UNKNOWN = 2;
            }
            Error error = 0;
        }
    }
}

message SatelliteMessage {
    oneof Message {
        message NotReady {} // this could be a grpc error rather than a message

        message TransferPiece {
            bytes piece_id; // the current piece-id
            bytes private_key;
            // addressed_order_limit contains the new piece id.
            AddressedOrderLimit addressed_order_limit;              
        }

        message DeletePiece {
            bytes piece_id;
        }

>         message Completed {

Reviewer: Should we have `Failed` message too?

Author: Updated

            // when everything is completed
            bytes exit_complete_signature;
        }
    }
}
```

> ## Open issues (if applicable)

Reviewer: My opinion is that all piece must be transferred regarding the durability level. The durability level should matter only to prioritize which pieces should be transferred first.

> ## Open issues (if applicable)

Reviewer 3: The issue is that when we take too much time to transfer pieces people might start hard-disconnecting, because it's not worth the escrow. This could end up storage nodes not bothering with graceful exit at all.

Ideally, yes, we would transfer all pieces.

> ## Open issues (if applicable)

Reviewer 4: In this case we should make sure to include the process in the design doc. We could have a configurable value `--graceful-exit.max-durability` or something that determines the cutoff where we decide we no longer care about transferring pieces on segments above a certain durability. Then we just configure it to `rs.Total` so that we transfer everything and we can lower it if we decide that takes too much time.

> ## Open issues (if applicable)

Author: Added as a future item in the overview.

When the upload has been verified the satellite will use `CompareAndSwap` to only switch the appropriate node without changing others. During the upload there may have been other things that happened:

- Segment / pointer was reuploaded. We'll ignore piece transferred for the exit.
- Segment / pointer pieces were changed, in that case we load the pointer, check whether the exiting storage node ID still needs to be replaced. If it does, then we retry `CompareAndSwap`. When the storage node at that piece number has changed, we'll ignore the piece transferred for the exit, because one of them needs to be discarded.
- Segment / pointer was deleted or is being deleted. We'll ignore the piece transferred for the exit.

> After changing the segment / pointer, we'll delete the piece from transfer queue, update progress, and send a `DeletePiece` to the exiting node.

Reviewer 2: In the cases above where we're ignoring the transferred piece, are we sending delete requests for those pieces?

Author: Agreed, Updated.
